In [1]:
# Update sklearn to prevent version mismatches
!pip install sklearn --upgrade

Requirement already up-to-date: sklearn in c:\users\budchris\anaconda3\envs\pythonadv\lib\site-packages (0.0)


In [2]:
# install joblib. This will be used to save your model. 
# Restart your kernel after installing 
!pip install joblib

In [3]:
import pandas as pd

# Read the CSV and Perform Basic Data Cleaning

In [4]:
df = pd.read_csv("exoplanet_data.csv")
# Drop the null columns where all values are null
df = df.dropna(axis='columns', how='all')
# Drop the null rows
df = df.dropna()
df.head()

koi_disposition  koi_fpflag_nt  koi_fpflag_ss  koi_fpflag_co  koi_fpflag_ec  \
0       CONFIRMED              0              0              0              0   
1  FALSE POSITIVE              0              1              0              0   
2  FALSE POSITIVE              0              1              0              0   
3       CONFIRMED              0              0              0              0   
4       CONFIRMED              0              0              0              0   

   koi_period  koi_period_err1  koi_period_err2  koi_time0bk  \
0   54.418383     2.479000e-04    -2.479000e-04   162.513840   
1   19.899140     1.490000e-05    -1.490000e-05   175.850252   
2    1.736952     2.630000e-07    -2.630000e-07   170.307565   
3    2.525592     3.760000e-06    -3.760000e-06   171.595550   
4    4.134435     1.050000e-05    -1.050000e-05   172.979370   

   koi_time0bk_err1  ...  koi_steff_err2  koi_slogg  koi_slogg_err1  \
0          0.003520  ...             -81      4.467           0.064   
1          0.000581  ...            -176      4.544           0.044   
2          0.000115  ...            -174      4.564           0.053   
3          0.001130  ...            -211      4.438           0.070   
4          0.001900  ...            -232      4.486           0.054   

   koi_slogg_err2  koi_srad  koi_srad_err1  koi_srad_err2         ra  \
0          -0.096     0.927          0.105         -0.061  291.93423   
1          -0.176     0.868          0.233         -0.078  297.00482   
2          -0.168     0.791          0.201         -0.067  285.53461   
3          -0.210     1.046          0.334         -0.133  288.75488   
4          -0.229     0.972          0.315         -0.105  296.28613   

         dec  koi_kepmag  
0  48.141651      15.347  
1  48.134129      15.436  
2  48.285210      15.597  
3  48.226200      15.509  
4  48.224670      15.714  

[5 rows x 41 columns]

# Select your features (columns)

In [5]:
# Set features. This will also be used as your x values.
# selected_features = df[['names', 'of', 'selected', 'features', 'here']]
selected_features = df

# Create a Train Test Split

Use `koi_disposition` for the y values

In [6]:
# Assign x and y values
X = df.drop("koi_disposition", axis=1)
y = df["koi_disposition"]

print(X.shape, y.shape)

(6991, 40) (6991,)


In [7]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, MinMaxScaler


# Split data into training and testing
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=115, stratify=y)

In [8]:
X_train.head()

koi_fpflag_nt  koi_fpflag_ss  koi_fpflag_co  koi_fpflag_ec  koi_period  \
2700              0              0              0              0   31.804760   
21                0              0              0              0    4.280964   
5040              0              1              1              0    0.829019   
2795              0              0              1              0    1.387838   
1765              0              0              0              0   14.383227   

      koi_period_err1  koi_period_err2  koi_time0bk  koi_time0bk_err1  \
2700         0.000091        -0.000091    190.25413           0.00224   
21           0.000006        -0.000006    171.89659           0.00115   
5040         0.000002        -0.000002    131.46667           0.00401   
2795         0.000006        -0.000006    132.11351           0.00318   
1765         0.000086        -0.000086    140.57648           0.00530   

      koi_time0bk_err2  ...  koi_steff_err2  koi_slogg  koi_slogg_err1  \
2700          -0.00224  ...            -128      4.235           0.143   
21            -0.00115  ...            -169      4.564           0.032   
5040          -0.00401  ...            -255      4.271           0.124   
2795          -0.00318  ...            -219      4.442           0.054   
1765          -0.00530  ...            -135      4.249           0.137   

      koi_slogg_err2  koi_srad  koi_srad_err1  koi_srad_err2         ra  \
2700          -0.117     1.356          0.212         -0.259  287.62082   
21            -0.168     0.831          0.207         -0.069  294.26581   
5040          -0.186     1.328          0.408         -0.220  296.03485   
2795          -0.216     1.044          0.335         -0.112  295.64212   
1765          -0.112     1.291          0.216         -0.216  290.08276   

            dec  koi_kepmag  
2700  43.141647      13.178  
21    49.314091      15.356  
5040  45.490822      14.661  
2795  48.773071      15.167  
1765  50.863480      13.312  

[5 rows x 40 columns]

# Pre-processing

Scale the data using the MinMaxScaler and perform some feature selection

In [9]:
# Scale your data
X_scaler = MinMaxScaler().fit(X_train)
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

# Train the Model



In [10]:
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier()
rf = rf.fit(X_train_scaled, y_train)

In [11]:
print(f"Training Data Score: {rf.score(X_train_scaled, y_train)}")
print(f"Testing Data Score: {rf.score(X_test_scaled, y_test)}")

Training Data Score: 1.0
Testing Data Score: 0.8907322654462243


In [12]:
importances = rf.feature_importances_
importances

array([0.09791249, 0.07294345, 0.10932576, 0.03378849, 0.02326187,
       0.01947587, 0.01709462, 0.01506121, 0.02050958, 0.02353691,
       0.01875072, 0.01063089, 0.01019446, 0.02456201, 0.03330046,
       0.03392097, 0.01862652, 0.01341758, 0.01227437, 0.04280815,
       0.0321966 , 0.0327231 , 0.0172136 , 0.0114898 , 0.01449425,
       0.01178813, 0.0602162 , 0.0026347 , 0.00947559, 0.03260869,
       0.03145577, 0.00942367, 0.00841646, 0.01121385, 0.00922148,
       0.01112481, 0.00874551, 0.01352694, 0.01075968, 0.00987482])

In [13]:
# Sort the features by their importance
sorted(zip(rf.feature_importances_, selected_features), reverse=True)

[(0.10932576343192908, 'koi_fpflag_ss'),
 (0.09791248827545278, 'koi_disposition'),
 (0.07294344987613344, 'koi_fpflag_nt'),
 (0.060216197285481166, 'koi_insol_err2'),
 (0.0428081520302859, 'koi_depth_err2'),
 (0.03392096941580573, 'koi_duration_err1'),
 (0.03378848555378621, 'koi_fpflag_co'),
 (0.03330045635155022, 'koi_duration'),
 (0.0327231038194, 'koi_prad_err1'),
 (0.03260869405565483, 'koi_steff'),
 (0.03219659919960611, 'koi_prad'),
 (0.03145576918776398, 'koi_steff_err1'),
 (0.024562005290344614, 'koi_impact_err2'),
 (0.023536907199359875, 'koi_time0bk_err1'),
 (0.023261873695557638, 'koi_fpflag_ec'),
 (0.02050957665382586, 'koi_time0bk'),
 (0.019475871721553576, 'koi_period'),
 (0.01875071602599881, 'koi_time0bk_err2'),
 (0.01862651814814427, 'koi_duration_err2'),
 (0.017213595333616865, 'koi_prad_err2'),
 (0.017094616934174446, 'koi_period_err1'),
 (0.01506120595086794, 'koi_period_err2'),
 (0.014494252402690805, 'koi_insol'),
 (0.013526935844567626, 'koi_srad_err2'),
 (0.01

# Hyperparameter Tuning

Use `GridSearchCV` to tune the model's parameters

In [14]:
# Create the GridSearchCV model
from sklearn.model_selection import GridSearchCV
param_grid = {'n_estimators': [250, 300, 350],
              'max_depth': [125, 150, 175]}
grid = GridSearchCV(rf, param_grid, verbose=3)

In [15]:
# Train the model with GridSearch
grid.fit(X_train_scaled, y_train)

Fitting 5 folds for each of 9 candidates, totalling 45 fits
[CV] max_depth=125, n_estimators=250 .................................


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


[CV] ..... max_depth=125, n_estimators=250, score=0.902, total=   7.4s
[CV] max_depth=125, n_estimators=250 .................................


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    7.3s remaining:    0.0s


[CV] ..... max_depth=125, n_estimators=250, score=0.888, total=   7.7s
[CV] max_depth=125, n_estimators=250 .................................


[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:   15.0s remaining:    0.0s


[CV] ..... max_depth=125, n_estimators=250, score=0.895, total=   7.6s
[CV] max_depth=125, n_estimators=250 .................................
[CV] ..... max_depth=125, n_estimators=250, score=0.893, total=   7.6s
[CV] max_depth=125, n_estimators=250 .................................
[CV] ..... max_depth=125, n_estimators=250, score=0.895, total=   7.8s
[CV] max_depth=125, n_estimators=300 .................................
[CV] ..... max_depth=125, n_estimators=300, score=0.901, total=   9.3s
[CV] max_depth=125, n_estimators=300 .................................
[CV] ..... max_depth=125, n_estimators=300, score=0.892, total=   9.2s
[CV] max_depth=125, n_estimators=300 .................................
[CV] ..... max_depth=125, n_estimators=300, score=0.898, total=   9.0s
[CV] max_depth=125, n_estimators=300 .................................
[CV] ..... max_depth=125, n_estimators=300, score=0.893, total=   9.2s
[CV] max_depth=125, n_estimators=300 .................................
[CV] .

[Parallel(n_jobs=1)]: Done  45 out of  45 | elapsed:  5.5min finished


GridSearchCV(estimator=RandomForestClassifier(),
             param_grid={'max_depth': [125, 150, 175],
                         'n_estimators': [250, 300, 350]},
             verbose=3)

In [16]:
print(grid.best_params_)
print(grid.best_score_)

{'max_depth': 175, 'n_estimators': 350}
0.8983402950101516


In [17]:
print(f"Training Grid Score: {grid.score(X_train_scaled, y_train)}")
print(f"Testing Grid Score: {grid.score(X_test_scaled, y_test)}")

Training Grid Score: 1.0
Testing Grid Score: 0.8947368421052632


In [18]:
# Make predictions with the hypertuned model
predictions = grid.predict(X_test_scaled)

In [19]:
# Calculate classification report
from sklearn.metrics import classification_report
print(classification_report(y_test, predictions))

                precision    recall  f1-score   support

     CANDIDATE       0.83      0.73      0.78       422
     CONFIRMED       0.78      0.86      0.82       450
FALSE POSITIVE       0.98      1.00      0.99       876

      accuracy                           0.89      1748
     macro avg       0.87      0.86      0.86      1748
  weighted avg       0.89      0.89      0.89      1748



# Save the Model

In [20]:
# save your model by updating "your_name" with your name
# and "your_model" with your model variable
# be sure to turn this in to BCS
# if joblib fails to import, try running the command to install in terminal/git-bash
import joblib
filename = 'solito_rf.sav'
joblib.dump(rf, filename)

['solito_rf.sav']

# Statements Regarding Model

Random Forest model is very conclusive in regards to identifying false positives (f1-score, FALSE POSITIVE = 0.99).

Hyperparameter tuning changed Testing Grid Score from 0.8907322654462243 to 0.8947368421052632